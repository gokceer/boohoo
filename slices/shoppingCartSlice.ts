import { createSlice } from "@reduxjs/toolkit";

import type { RootState } from '../store'
import type { CommonProductObjectType } from "../App"

type initialStateType = {
  items: CommonProductObjectType[]
}

const initialState: initialStateType = {
  items: [],
};

export const shoppingCartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addToCart: (state, action) => {
      state.items = [...state.items, action.payload];
    },
    removeFromCart: (state, action) => {
      const index = state.items.findIndex((item) => item.id === action.payload.id);
        
      let tempCart = [...state.items];
        
      index >= 0
        ? tempCart.splice(index, 1)
        : console.warn(`Cannot remove id: ${action.payload.id}`);
        
      state.items = tempCart;
    },
    removeAllFromCartWithId: (state, action) => {
      const filteredState = state.items.filter((item) => {
        return item.id !== action.payload.id
      })
      state.items = [...filteredState]
    },
  }
});
  
  export const { addToCart, removeFromCart, removeAllFromCartWithId } = shoppingCartSlice.actions;
  
  export const selectCartItems = (state: RootState) => state.cart.items;
  
  export const selectCartItemsWithId = (state: RootState, id: number) =>
  state.cart.items.filter((item) => item.id === id);
  
  export const selectCartTotal = (state: RootState) =>
  state.cart.items.reduce((total, item) => (total += item.price), 0);
  
  export default shoppingCartSlice.reducer;
