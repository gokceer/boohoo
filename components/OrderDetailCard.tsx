import React from 'react'
import { View, Text, Button, StyleSheet } from 'react-native'
import { useAppSelector } from '../hooks'
import { selectCartTotal } from "../slices/shoppingCartSlice"

const OrderDetailCard = () => {
  // Take total price in the cart
  const cartTotal = useAppSelector(selectCartTotal);

  // Display subtotal, delivery fee and total price in a card.
  return (
    <>
      {cartTotal > 0 &&
        <View style={styles.container}>
          <View style={styles.textRows}>
            <Text>Subtotal</Text>
            <Text>£ {cartTotal}</Text>
          </View>

          <View style={styles.textRows}>
            <Text>Delivery Fee</Text>
            <Text>£ {5}</Text>
          </View>

          <View style={styles.textRows}>
            <Text style={styles.price}>Order Total </Text>
            <Text style={styles.price}>£ {cartTotal + 5}</Text>
          </View>

          <Button
            onPress={null}
            title="Place Order"
            color="#841584"
            accessibilityLabel="Place Order Button"
          />
        </View>}
    </>
  )
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white"
  },
  textRows: {
    flexDirection: "row",
    justifyContent: "space-between",
    paddingHorizontal: 12,
    paddingVertical: 7
  },
  price: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 8
  }
})

export default OrderDetailCard