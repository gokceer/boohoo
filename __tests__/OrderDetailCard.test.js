import { Provider } from "react-redux";
import { store } from "../store";
import { render, act } from "@testing-library/react-native";
import OrderDetailCard from "../components/OrderDetailCard";
import { addToCart } from "../slices/shoppingCartSlice";

const mockItemProp = {
  id: 1,
  colour: "Black",
  name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
  price: 10,
  img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
};

const setup = (props = {}) => {
  const wrapper = render(
    <Provider store={store}>
      <OrderDetailCard {...props} />
    </Provider>
  );
  return wrapper;
};

describe("Order Detail unit tests", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("renders correctly, snapshot test", () => {
    const tree = setup();
    expect(tree).toMatchSnapshot();
  });

  it("Empty cart OrderDetailCard not exists test", () => {
    const wrapper = setup();
    expect(wrapper.queryByText("Subtotal")).not.toBeTruthy();
    expect(wrapper.queryByText("Delivery Fee")).not.toBeTruthy();
    expect(wrapper.queryByText("Order Total")).not.toBeTruthy();
  });

  it("OrderDetailCard existance test", () => {
    const wrapper = setup();

    // add item to the store
    act(() => {
      store.dispatch(addToCart(mockItemProp));
    });

    expect(wrapper.queryByText("Subtotal")).toBeTruthy();
    expect(wrapper.queryByText("£ 10")).toBeTruthy();

    expect(wrapper.queryByText("Delivery Fee")).toBeTruthy();
    expect(wrapper.queryByText("£ 5")).toBeTruthy();

    expect(wrapper.queryByText("Order Total")).toBeTruthy();
    expect(wrapper.queryByText("£ 15")).toBeTruthy();
  });
});
