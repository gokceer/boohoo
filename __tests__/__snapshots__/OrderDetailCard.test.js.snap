// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Detail unit tests renders correctly, snapshot test 1`] = `null`;
